platform/platform-impl/src/com/intellij/openapi/vfs/impl/local/LocalFileSystemBase.java

```diff
@@ -517,6 +517,22 @@ public boolean markNewFilesAsDirty() {
       if (SystemInfo.isWindows) {
         return convertToNioFileAndCheck(file, false).toRealPath(LinkOption.NOFOLLOW_LINKS).getFileName().toString();
       }
+      else {
+        // Not Windows (i.e. probably Unix, probably probably MacOS X with its case-insensitive FS).
+        //
+        // On OS X, we would like to query the filesystem directly for metadata, using an interface to the
+        // getattrlist() system call, which would return directly the metadata associated with the existing file, including
+        // its filesystem name.  However, we don't have an interface to that in the JVM as far as I am aware.
+        //
+        // Instead, handle one common case as quickly as possible: compute the file's realpath, resolving links
+        // (to avoid quadratic behaviour from directory listing).  In general this is not a suitable canonical name for the
+        // file, because links have been resolved, but if the result compares with case-insensitive equality with the given
+        // file's path, then the return value is suitable for use as a canonical name.
+        var resolvedRealPath = nioFile.toRealPath();
```

Comment:
I think you want to add `Path nioFile = convertToNioFileAndCheck(file, false);` on line 516 for the code to compile.
Then you can replace the calls to `convertToNioPath...` on lines 518 and 537 with `nioFile`.

Reply:
Heh, yes, thank you.  (I did warn you that I hadn't tried compiling it...).  Updated to something hopefully a bit more correct.  (Not the one on line 537 which gets the nio file for `parent`).